File: 백준/Silver/11899. 괄호 끼워넣기/괄호 끼워넣기.cc
#include<iostream>
#include<string>
#include<stack>
using namespace std;

int main()
{
	ios::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);

	string s;
	cin >> s;

	stack<int> st;
	for (int i = 0; i < s.length(); i++) {
		if (st.empty()) {
			st.push(s[i]);
			continue;
		}
		else if (st.top() == '(') {
			if (s[i] == '(') st.push(s[i]);
			else st.pop();
		}
		else if (st.top() == ')') st.push(s[i]);
	}
	cout << st.size();
	return 0;
}

File: 백준/Silver/11899. 괄호 끼워넣기/괄호 끼워넣기.java
import java.util.*;
import java.io.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String s = br.readLine();

		Stack<Character> st = new Stack<>();
		for (int i = 0; i < s.length(); i++) {
			if (st.empty()) {
				st.push(s.charAt(i));
				continue;
			} else if (st.peek() == '(') {
				if(s.charAt(i)=='(') st.push(s.charAt(i));
				else st.pop();
			}
			else if(st.peek()==')') st.push(s.charAt(i));
		}
		System.out.println(st.size());
	}
}
